### PR TITLE
[SPARK-36703][SQL] Remove the Sort if it is the child of RepartitionByExpression

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr removes the `Sort` if it is the child of `RepartitionByExpression`. For example:
```sql
spark.range(10).selectExpr("cast(id AS string)").sort("id").repartition($"id").explain()
```
Before this PR:
```
== Optimized Logical Plan ==
RepartitionByExpression [id#2]
+- Sort [id#2 ASC NULLS FIRST], true
   +- Project [cast(id#0L as string) AS id#2]
      +- Range (0, 10, step=1, splits=Some(2))
```
After this PR:
```
== Optimized Logical Plan ==
RepartitionByExpression [id#2]
+- Project [cast(id#0L as string) AS id#2]
   +- Range (0, 10, step=1, splits=Some(2))
```

### Why are the changes needed?

Remove useless sort. 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -987,10 +987,10 @@ object CollapseRepartition extends Rule[LogicalPlan] {
       case (false, true) => if (r.numPartitions >= child.numPartitions) child else r
       case _ => r.copy(child = child.child)
     }
-    // Case 2: When a RepartitionByExpression has a child of Repartition or RepartitionByExpression
-    // we can remove the child.
-    case r @ RepartitionByExpression(_, child: RepartitionOperation, _) =>
-      r.copy(child = child.child)
+    // Case 2: When a RepartitionByExpression has a child of Sort, Repartition or
+    // RepartitionByExpression we can remove the child.
+    case r @ RepartitionByExpression(_, child @ (_: Sort | _: RepartitionOperation), _) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:992]: is it possible that Spark can keep the ordering in some cases?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:992]: e.g. partition key is the same as sort key?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:992]: Do you mean `spark.range(10).selectExpr("cast(id AS string)").repartition($"id").sort("id")`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:992]: e.g. `spark.table(...).repartition("a").sortWithinPartitions("a").repartition(a)`.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: ok. make it only remove the global sort.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
also cc @sigmod

**Comment by sigmod** [comment]
> also cc @sigmod

LGTM!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**